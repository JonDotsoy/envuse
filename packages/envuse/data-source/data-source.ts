import { Base } from "./statements/components/base";
import { Block, BlockType } from "./statements/components/block";
import { ArrCursor } from "./statements/lib/arr-cursor";
import { CommentOperator } from "./statements/components/comment-operator";
import { UnexpectedTokenError } from "./statements/tdo/unexpected-token-error";
import { Variable } from "./statements/components/variable";
import { StatementObject } from "./statements/components/statement-object";
import fs from "fs";
import { BlockComment } from "./statements/components/block-comment";
import { CommentInline } from "./statements/components/comment-inline";
import { stringify as DataSourceStringify } from "./statements/lib/stringify";
import { StringifyOptions } from "./statements/lib/stringify-options";

// data source

export type CustomType = {
  type: string;
  parser: (ctx: {
    valueStr: string;
    elementVariable: Variable;
    elementDescription: null | BlockComment;
  }) => any;
};

const stringType: CustomType = {
  type: "string",
  parser: (ctx) => {
    const valueStr = ctx.valueStr;
    return valueStr;
  },
};

const numberType: CustomType = {
  type: "number",
  parser: (ctx) => {
    const valueStr = ctx.valueStr;
    const v = Number(valueStr);
    if (isNaN(v) || v === Infinity) {
      ctx.elementVariable.rejectUnexpectedTokenError({
        message: `${valueStr} is not a valid number`,
        position: ctx.elementVariable.valueVariable.pos,
      });
    }
    return v;
  },
};

const booleanType: CustomType = {
  type: "boolean",
  parser: (ctx) => {
    const valueStr = ctx.valueStr;
    if (["true", "false"].indexOf(valueStr) === -1) {
      ctx.elementVariable.rejectUnexpectedTokenError({
        message: `${valueStr} is not a valid boolean`,
        position: ctx.elementVariable.valueVariable.pos,
      });
    }
    return valueStr === "true";
  },
};

export const defaultCustomTypes: CustomType[] = [
  stringType,
  numberType,
  booleanType,
];

export type Definition = {
  type: string;
  description: string | null;
  value: any;
  valueStr: string;
  elementVariable: Variable;
  elementDescription?: BlockComment | null;
};

export type Option =
  | Buffer
  | {
      filename?: string | null;
      body: Buffer;
    };

export type Values = {
  [k: string]: any;
};

export type CompileOptions = {
  customTypes?: CustomType[];
};

/** AST Parser */
export class DataSource {
  private constructor(private filename: string | null, private body: Buffer) {}

  toAstBody() {
    try {
      return Base.createElement(new Block(this.filename, this.body, 0));
    } catch (err) {
      if (UnexpectedTokenError.isUnexpectedTokenError(err)) {
        // Error.captureStackTrace(err, EnvuseFileParser.prototype.toAstBody);
        // const stack = err.stack;
        // if (stack) {
        //   // let posEndMessage = err.name.length + err.message.length + 3;
        // }
      }
      throw err;
    }
  }

  private static makeOperatorEvaluator(operator: CommentOperator) {
    return (value: { [k: string]: any }) => {
      const a = new ArrCursor(operator.statement.statements);
      let accumulator: any;

      const getValue = (statementObject: StatementObject) => {
        // console.log(inspect(statementObject));
        switch (statementObject.type) {
          case "NameInstance":
            return (
              Array.isArray(statementObject.value) &&
              statementObject.value.reduce(
                (r, path) => (typeof r === "object" ? r[path] : undefined),
                value
              )
            );
          case "Number":
          case "Boolean":
          case "String":
            return statementObject.value;
          default:
            throw new Error("unprocessable statement");
        }
      };

      while (a.has()) {
        if (a.position === 0) {
          accumulator = getValue(a.current());
          a.forward();
          continue;
        }

        switch (a.current().type) {
          case "StrictEqualitySymbol": {
            a.forward();
            // console.log({ accumulator, currentType: a.current().type, currentValue: a.current().value, parsed: getValue(a.current()), value })
            accumulator = accumulator === getValue(a.current());
            a.forward();
            break;
          }
          default:
            throw new Error(`Statement no expected`);
        }
      }

      return accumulator;
    };
  }

  static parse(options: Option & CompileOptions, values?: Values) {
    const ast = this.createDataSource(options);

    const operatorsList = ast.elementList
      .filter(
        (element): element is CommentOperator =>
          element instanceof CommentOperator
      )
      .map((operator) => ({
        operator,
        assert: this.makeOperatorEvaluator(operator),
      }));

    const isVariable = (element: Base): element is Variable =>
      element instanceof Variable;
    const isDescriptor = (element: Base): element is BlockComment =>
      element instanceof BlockComment;
    const isComment = (element: Base): element is CommentInline =>
      element instanceof CommentInline;

    // const variableList = ast.elementList.filter(
    //   (element): element is Variable => element instanceof Variable
    // );

    let description: BlockComment | null = null;

    const variableList = ast.elementList.reduce((acum, element) => {
      if (isDescriptor(element)) {
        description = element;
      }
      if (isComment(element)) {
        description = null;
      }
      if (isVariable(element)) {
        acum.push([element, description]);
        description = null;
      }
      return acum;
    }, [] as [Variable, BlockComment | null][]);

    const definitions = variableList.reduce((acum, [element, description]) => {
      const operators = operatorsList.filter((operator) =>
        operator.operator.elementList.includes(element)
      );

      if (operators.length) {
        const result = operators.reduce(
          (v, operator) =>
            v &&
            operator.assert({
              ...values,
              ...Object.fromEntries(
                Object.entries(acum).map(([k, v]) => [k, v.value])
              ),
            }),
          true
        );
        if (!result) {
          return acum;
        }
      }

      const type = element.typeVariable?.value ?? "string";
      const valueStr = element.valueVariable.value;

      const customTypes = [
        ...defaultCustomTypes,
        ...(options.customTypes ?? []),
      ];

      const getValue = () => {
        for (const customType of customTypes) {
          if (customType.type === type) {
            return customType.parser({
              elementDescription: description,
              elementVariable: element,
              valueStr,
            });
          }
        }

        throw new Error(`Unsupported type ${type}`);
      };

      return {
        ...acum,
        [element.keyVariable.value]: {
          type,
          valueStr,
          description: description?.value ?? null,
          value: getValue(),
          elementVariable: element,
          elementDescription: description,
        },
      };
    }, {} as { [k: string]: Definition });

    const parsed = Object.entries(definitions).reduce(
      (acum, [name, def]) => ({ ...acum, [name]: def.valueStr }),
      {} as { [k: string]: string }
    );

    return { definitions, parsed, ast } as const;
  }

  static parseFile(
    filename: string,
    values?: { [k: string]: any },
    options?: CompileOptions
  ) {
    // read file and store buffer
    const buffer = fs.readFileSync(filename);

    return this.parse({ filename, body: buffer, ...options }, values);
  }

  static createDataSource(options: Option) {
    if (options instanceof Buffer) {
      return new DataSource(null, options).toAstBody();
    }
    return new DataSource(options.filename ?? null, options.body).toAstBody();
  }

  static stringify(comp: BlockType, options?: StringifyOptions) {
    return DataSourceStringify(comp, options);
  }
}
