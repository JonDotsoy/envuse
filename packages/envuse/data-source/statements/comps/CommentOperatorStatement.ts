import { Base } from "./Base";
import { BufferCursor } from "../lib/BufferCursor";
import { BCharType } from "../tdo/BCharType";
import { Space } from "./Space";
import { CharactersKey as K } from "../tdo/CharactersKey";
import { StatementObject, StatementObjectType } from "./StatementObject";
import { BaseSerializeOption } from "../tdo/BaseSerializeOption";

export type CommentOperatorStatementType = {
  $type: "CommentOperatorStatement";
  statements: StatementObjectType[];
};

export class CommentOperatorStatement extends Base {
  $type = "CommentOperatorStatement" as const;
  statements: StatementObject[] = [];

  prepare(bufferCursor: BufferCursor<BCharType>): void {
    while (true) {
      if (
        bufferCursor.current() === K.newLineLF ||
        bufferCursor.current() === K.numberSign ||
        bufferCursor.isClosed()
      ) {
        bufferCursor.forward();
        return;
      }

      this.statements.push(this.createElement(StatementObject));

      if (bufferCursor.current() === K.newLineLF) {
        bufferCursor.forward();
        return;
      }

      if (bufferCursor.current() === K.space) {
        this.createElement(Space);
      }
    }
  }

  toJSON() {
    return {
      ...super.toJSON(),
      statements: this.statements,
    };
  }

  static serialize(comp: CommentOperatorStatementType) {
    return Buffer.concat(
      comp.statements.map((s, i, { length }) =>
        Buffer.concat([
          StatementObject.serialize(s),
          Buffer.from(length - 1 === i ? "" : " "),
        ])
      )
    );
  }
}
