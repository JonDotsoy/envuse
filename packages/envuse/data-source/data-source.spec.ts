import { DataSource } from "./data-source";
import { takeDemoFile } from "./statements/lib/takeDemoFile";
import util from "util";
import { CommentOperator } from "./statements/comps/CommentOperator";
import { CommentOperatorStatement } from "./statements/comps/CommentOperatorStatement";
import { StatementObject } from "./statements/comps/StatementObject";
import { Base } from "./statements/comps/Base";
import { b } from "./statements/lib/toBuffer";

describe("DataSource", () => {
  it("shoud make a ast out", () => {
    const [filename1, body1] = takeDemoFile();

    const envuseFileParser = new DataSource(filename1, body1);
    expect(
      util.formatWithOptions(
        { depth: Infinity },
        "",
        envuseFileParser.toAstBody()
      )
    ).toMatchSnapshot();
  });

  it("shoud make a ast with operators", () => {
    const [filename, body] = takeDemoFile();

    const envuseFileParser = new DataSource(filename, body);
    expect(
      util.formatWithOptions(
        { depth: Infinity },
        "",
        envuseFileParser.toAstBody()
      )
    ).toMatchSnapshot();
    // console.log(util.formatWithOptions({ depth: Infinity }, '', envuseFileParser.toAstBody()))
  });

  it("shoud make a ast with multiple blocks", () => {
    const [filename, body] = takeDemoFile(".env");

    const envuseFileParser = new DataSource(filename, body);
    expect(
      util.formatWithOptions(
        { depth: Infinity },
        "",
        envuseFileParser.toAstBody()
      )
    ).toMatchSnapshot();
    // console.log(util.formatWithOptions({ depth: Infinity }, '', envuseFileParser.toAstBody()))
  });

  describe("parse statement", () => {
    it("parse statement operation", () => {
      const body = Buffer.from("true ===   false # asd");

      const cmp = Base.createElement(
        new CommentOperatorStatement("a", body, 0)
      );

      expect(cmp).toMatchSnapshot();
    });

    it("should parse statement with numbers", () => {
      const body = Buffer.from("34214324 ===   22_343.32 # comment");

      const cmp = Base.createElement(
        new CommentOperatorStatement("a", body, 0)
      );

      expect(cmp).toMatchSnapshot();
      // console.log(cmp)
    });

    it("should parse StatementObject string", () => {
      const [fl, body] = takeDemoFile();

      const cmp = Base.createElement(new StatementObject(fl, body, 0));

      // expect(cmp).toMatchSnapshot()
      // console.log(cmp)
    });

    it("should parse statement with string", () => {
      const [fl, body] = takeDemoFile();

      Base.createElement(new CommentOperatorStatement(fl, body, 0));

      // expect(cmp).toMatchSnapshot()
      // console.log(cmp)
    });

    it("should parse statement with string and special characters", () => {
      const [fl, body] = takeDemoFile(".special-characters.txt");

      const cmp = Base.createElement(new StatementObject(fl, body, 0));

      expect(cmp.value).toEqual("\n\t");
    });

    it("should parse instance path", () => {
      const body = Buffer.from("obj.a.b.VAL2");

      const cmp = Base.createElement(new StatementObject("a", body, 0));

      expect(cmp).toBeInstanceOf(StatementObject);
      expect(cmp.$type).toEqual("StatementObject");
      expect(cmp.type).toEqual("NameInstance");
      expect(cmp.value).toEqual(["obj", "a", "b", "VAL2"]);
    });

    it("should parse numbers", () => {
      const body = b("1_321.123");

      const cmp = Base.createElement(new StatementObject("a", body, 0));

      expect(cmp.value).toEqual(1_321.123);
    });

    it("should parse strings", () => {
      const body = b('"hi"');

      const cmp = Base.createElement(new StatementObject(null, body, 0));

      expect(cmp.value).toEqual("hi");
    });

    it("should parse statement with variables", () => {
      const body = Buffer.from("VAL1 ===   obj.a.b.VAL2 # comment");

      const cmp = Base.createElement(
        new CommentOperatorStatement("a", body, 0)
      );

      expect(cmp.elementList.map((e) => e.toString())).toMatchInlineSnapshot(`
        Array [
          "CommentOperatorStatement (0, 25): \\"VAL1 ===   obj.a.b.VAL2 #\\"",
          "StatementObject<NameInstance> (0, 4): \\"VAL1\\"",
          "Space (4, 5): \\" \\"",
          "StatementObject<StrictEqualitySymbol> (5, 8): \\"===\\"",
          "Space (8, 11): \\"   \\"",
          "StatementObject<NameInstance> (11, 23): \\"obj.a.b.VAL2\\"",
          "Space (23, 24): \\" \\"",
        ]
      `);
    });

    it("should parse end parsing comment code", () => {
      const body = b("#! if true\nFOO=bar");

      const cmp = Base.createElement(new CommentOperator("a", body, 0));

      // expect(cmp).toMatchInlineSnapshot(`Object {}`)
      expect(cmp.operator.raw.toString()).toEqual("if");
      // console.log(cmp)
    });

    it("should read property list children", () => {
      const [fl, body] = takeDemoFile();

      const envuseFileParser = new DataSource(fl, body);

      const b = envuseFileParser.toAstBody();

      for (const elem of b.elementList) {
        expect(elem).toBeInstanceOf(Base);
      }

      expect(
        b.elementList.map((e) => `${e.constructor.name} <${e.pos}, ${e.end}>`)
      ).toMatchInlineSnapshot(`
        Array [
          "Block <0, 35>",
          "Variable <0, 8>",
          "VariableKey <0, 3>",
          "SymbolEqual <3, 4>",
          "VariableValue <4, 8>",
          "CommentOperator <8, 34>",
          "Space <10, 11>",
          "VariableKey <11, 13>",
          "Space <13, 14>",
          "CommentOperatorStatement <14, 19>",
          "StatementObject <14, 18>",
          "Block <19, 34>",
          "SpaceNewLine <19, 20>",
          "Variable <20, 28>",
          "VariableKey <20, 23>",
          "SymbolEqual <23, 24>",
          "VariableValue <24, 28>",
          "SpaceNewLine <28, 29>",
          "CommentOperator <29, 34>",
          "Space <31, 32>",
          "VariableKey <32, 34>",
          "SpaceNewLine <34, 35>",
        ]
      `);
    });

    it("should parse comment operator correctly", () => {
      expect(
        DataSource.createDataSource({
          filename: null,
          body: Buffer.from(`#;if a.b.c.d ===\nfoo="bar"\n#;fi\n`),
        }).elementList.map((e) => e.toString())
      ).toMatchSnapshot();
      expect(
        DataSource.createDataSource({
          filename: null,
          body: Buffer.from(`#;if true\nfoo="bar"\n#;fi\n`),
        }).elementList.map((e) => e.toString())
      ).toMatchSnapshot();
      expect(
        DataSource.createDataSource({
          filename: null,
          body: Buffer.from(`#;if 123\nfoo="bar"\n#;fi\n`),
        }).elementList.map((e) => e.toString())
      ).toMatchSnapshot();
      expect(
        DataSource.createDataSource({
          filename: null,
          body: Buffer.from(`#;if var\nfoo="bar"\n#;fi\n`),
        }).elementList.map((e) => e.toString())
      ).toMatchSnapshot();
      expect(
        DataSource.createDataSource({
          filename: null,
          body: Buffer.from(`#;if foo === bar\nfoo="bar"\n#;fi\n`),
        }).elementList.map((e) => e.toString())
      ).toMatchSnapshot();
    });

    it("should parse block descriptive comment", () => {
      const [fl, demo] = takeDemoFile()

      const envuseFileParser = DataSource.createDataSource({ filename: fl, body: demo });

      expect(envuseFileParser).toMatchInlineSnapshot();
    });
  });
});

describe("DataSource file parse", () => {
  it("parse file", () => {
    const res = DataSource.parse({ body: b(`foo=var\nabc=def\naaa=32`) });

    expect(res.parsed).toEqual({
      foo: "var",
      abc: "def",
      aaa: "32",
    });
  });

  it("should parse with mixing conditionals", () => {
    const res1 = DataSource.parse({
      body: b(`foo=var\nabc=def\naaa=32\n#; if true\nbbb=ccc\n#; fi`),
    });

    expect(res1.parsed).toEqual({
      foo: "var",
      abc: "def",
      aaa: "32",
      bbb: "ccc",
    });

    const res2 = DataSource.parse({
      body: b(`foo=var\nabc=def\naaa=32\n#; if false\nbbb=ccc\n#; fi`),
    });

    expect(res2.parsed).toEqual({
      foo: "var",
      abc: "def",
      aaa: "32",
    });

    const res3 = DataSource.parse({
      body: b(
        `aaa=aaa\n#; if true\nbbb=bbb\n#; if false\nccc=ccc\n#; fi\n#; fi`
      ),
    });

    expect(res3.parsed).toEqual({
      aaa: "aaa",
      bbb: "bbb",
    });

    const res4 = DataSource.parse({
      body: b(
        `aaa=aaa\n#; if false\nbbb=bbb\n#; if true\nccc=ccc\n#; fi\n#; fi`
      ),
    });

    expect(res4.parsed).toEqual({
      aaa: "aaa",
    });
  });

  describe("complex sentences", () => {
    it("should parse with complex sentences (booleans)", () => {
      const res = DataSource.parse({
        body: b(`#; if true === true\naaa=aaa\n#; fi`),
      });

      expect(res.parsed).toStrictEqual({
        aaa: "aaa",
      });
    });

    it("should parse with complex sentences (booleans)", () => {
      const res = DataSource.parse({
        body: b(`bbb=bbb\n#; if true === true === false\naaa=aaa\n#; fi`),
      });

      expect(res.parsed).toStrictEqual({
        bbb: "bbb",
      });
    });

    it("should parse with complex sentences (strings)", () => {
      const res = DataSource.parse({
        body: b(`aaa=aaa\n#; if 'aaa' === 'aaa'\nbbb=bbb\n#; fi`),
      });

      expect(res.parsed).toStrictEqual({
        aaa: "aaa",
        bbb: "bbb",
      });
    });

    it("should parse with complex sentences (strings)", () => {
      const res = DataSource.parse({
        body: b(`aaa=aaa\n#; if 'ccc' === 'aaa'\nbbb=bbb\n#; fi`),
      });

      expect(res.parsed).toStrictEqual({
        aaa: "aaa",
      });
    });

    it("should parse with complex sentences (booleans)", () => {
      const res = DataSource.parse({
        body: b(`aaa=aaa\n#; if 12 === 12\nbbb=bbb\n#; fi`),
      });

      expect(res.parsed).toStrictEqual({
        aaa: "aaa",
        bbb: "bbb",
      });
    });

    it("should parse with complex sentences (booleans)", () => {
      const res = DataSource.parse({
        body: b(`aaa=aaa\n#; if 13_3 === 12\nbbb=bbb\n#; fi`),
      });

      expect(res.parsed).toStrictEqual({
        aaa: "aaa",
      });
    });

    it("should parse with complex sentences (name instance)", () => {
      const res = DataSource.parse({
        body: b(`aaa=aaa\n#; if 'aaa' === aaa\nbbb=bbb\n#; fi`),
      });

      expect(res.parsed).toStrictEqual({
        aaa: "aaa",
        bbb: "bbb",
      });
    });

    it("should parse with complex sentences (name instance)", () => {
      const res = DataSource.parse({
        body: b(`aaa=aaa\n#; if 'ccc' === aaa\nbbb=bbb\n#; fi`),
      });

      expect(res.parsed).toStrictEqual({
        aaa: "aaa",
      });
    });

    it("should parse with complex sentences (name instance and values)", () => {
      const res = DataSource.parse(
        { body: b(`aaa=aaa\n#; if 'ccc' === externalVal\nbbb=bbb\n#; fi`) },
        { externalVal: "ccc" }
      );

      expect(res.parsed).toStrictEqual({
        aaa: "aaa",
        bbb: "bbb",
      });
    });
  });
});

describe("DataSource file stringify", () => {
  it("should stringify ast", () => {
    const ast = DataSource.createDataSource(b(`foo=bar`));

    const str = JSON.stringify(ast, null, 2);

    expect(str).toMatchInlineSnapshot(`
      "{
        \\"$type\\": \\"Block\\",
        \\"pos\\": 0,
        \\"end\\": 7,
        \\"children\\": [
          {
            \\"$type\\": \\"Variable\\",
            \\"pos\\": 0,
            \\"end\\": 7,
            \\"keyVariable\\": {
              \\"$type\\": \\"VariableKey\\",
              \\"pos\\": 0,
              \\"end\\": 3,
              \\"value\\": \\"foo\\"
            },
            \\"valueVariable\\": {
              \\"$type\\": \\"VariableValue\\",
              \\"pos\\": 4,
              \\"end\\": 7,
              \\"value\\": \\"bar\\"
            }
          }
        ]
      }"
    `);
  });

  it("should stringify ast", () => {
    // const ast = EnvuseFileParser.parseToAst(b('a=b\n#; if true\nc=d\n#; fi'))
    const ast = DataSource.createDataSource(b("#; if true\n#; fi"));

    const str = JSON.stringify(ast, null, 2);

    expect(str).toMatchInlineSnapshot(`
      "{
        \\"$type\\": \\"Block\\",
        \\"pos\\": 0,
        \\"end\\": 16,
        \\"children\\": [
          {
            \\"$type\\": \\"CommentOperator\\",
            \\"pos\\": 0,
            \\"end\\": 16,
            \\"operator\\": {
              \\"$type\\": \\"VariableKey\\",
              \\"pos\\": 3,
              \\"end\\": 5,
              \\"value\\": \\"if\\"
            },
            \\"statement\\": {
              \\"$type\\": \\"CommentOperatorStatement\\",
              \\"pos\\": 6,
              \\"end\\": 11,
              \\"statements\\": [
                {
                  \\"$type\\": \\"StatementObject\\",
                  \\"pos\\": 6,
                  \\"end\\": 10,
                  \\"type\\": \\"Boolean\\",
                  \\"value\\": true
                }
              ]
            },
            \\"block\\": {
              \\"$type\\": \\"Block\\",
              \\"pos\\": 11,
              \\"end\\": 16,
              \\"children\\": [
                {
                  \\"$type\\": \\"CommentOperator\\",
                  \\"pos\\": 11,
                  \\"end\\": 16,
                  \\"operator\\": {
                    \\"$type\\": \\"VariableKey\\",
                    \\"pos\\": 14,
                    \\"end\\": 16,
                    \\"value\\": \\"fi\\"
                  }
                }
              ]
            }
          }
        ]
      }"
    `);
  });
});
