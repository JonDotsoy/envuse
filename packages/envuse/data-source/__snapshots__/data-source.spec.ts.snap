// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSource parse statement parse statement operation 1`] = `
Object {
  "$type": "CommentOperatorStatement",
  "children": undefined,
  "end": 18,
  "pos": 0,
  "statements": Array [
    Object {
      "$type": "StatementObject",
      "children": undefined,
      "end": 4,
      "pos": 0,
      "type": "Boolean",
      "value": true,
    },
    Object {
      "$type": "StatementObject",
      "children": undefined,
      "end": 8,
      "pos": 5,
      "type": "StrictEqualitySymbol",
      "value": undefined,
    },
    Object {
      "$type": "StatementObject",
      "children": undefined,
      "end": 16,
      "pos": 11,
      "type": "Boolean",
      "value": false,
    },
  ],
}
`;

exports[`DataSource parse statement should parse comment operator correctly 1`] = `
Array [
  "Block (0, 32): \\"#;if a.b....",
  "CommentOperator (0, 31): \\"#;if a.b....",
  "VariableKey (2, 4): \\"if\\"",
  "Space (4, 5): \\" \\"",
  "CommentOperatorStatement (5, 17): \\"a.b.c.d =...",
  "StatementObject<NameInstance> (5, 12): \\"a.b.c.d\\"",
  "Space (12, 13): \\" \\"",
  "StatementObject<StrictEqualitySymbol> (13, 16): \\"===\\"",
  "Block (17, 31): \\"foo=\\\\\\"bar...",
  "Variable (17, 27): \\"foo=\\\\\\"bar...",
  "VariableKey (17, 20): \\"foo\\"",
  "SymbolEqual (20, 21): \\"=\\"",
  "VariableValue (21, 26): \\"\\\\\\"bar\\\\\\"\\"",
  "CommentOperator (27, 31): \\"#;fi\\"",
  "VariableKey (29, 31): \\"fi\\"",
  "SpaceNewLine (31, 32): \\"\\\\n\\"",
]
`;

exports[`DataSource parse statement should parse comment operator correctly 2`] = `
Array [
  "Block (0, 25): \\"#;if true...",
  "CommentOperator (0, 24): \\"#;if true...",
  "VariableKey (2, 4): \\"if\\"",
  "Space (4, 5): \\" \\"",
  "CommentOperatorStatement (5, 10): \\"true\\\\n\\"",
  "StatementObject<Boolean> (5, 9): \\"true\\"",
  "Block (10, 24): \\"foo=\\\\\\"bar...",
  "Variable (10, 20): \\"foo=\\\\\\"bar...",
  "VariableKey (10, 13): \\"foo\\"",
  "SymbolEqual (13, 14): \\"=\\"",
  "VariableValue (14, 19): \\"\\\\\\"bar\\\\\\"\\"",
  "CommentOperator (20, 24): \\"#;fi\\"",
  "VariableKey (22, 24): \\"fi\\"",
  "SpaceNewLine (24, 25): \\"\\\\n\\"",
]
`;

exports[`DataSource parse statement should parse comment operator correctly 3`] = `
Array [
  "Block (0, 24): \\"#;if 123\\\\...",
  "CommentOperator (0, 23): \\"#;if 123\\\\...",
  "VariableKey (2, 4): \\"if\\"",
  "Space (4, 5): \\" \\"",
  "CommentOperatorStatement (5, 9): \\"123\\\\n\\"",
  "StatementObject<Number> (5, 8): \\"123\\"",
  "Block (9, 23): \\"foo=\\\\\\"bar...",
  "Variable (9, 19): \\"foo=\\\\\\"bar...",
  "VariableKey (9, 12): \\"foo\\"",
  "SymbolEqual (12, 13): \\"=\\"",
  "VariableValue (13, 18): \\"\\\\\\"bar\\\\\\"\\"",
  "CommentOperator (19, 23): \\"#;fi\\"",
  "VariableKey (21, 23): \\"fi\\"",
  "SpaceNewLine (23, 24): \\"\\\\n\\"",
]
`;

exports[`DataSource parse statement should parse comment operator correctly 4`] = `
Array [
  "Block (0, 24): \\"#;if var\\\\...",
  "CommentOperator (0, 23): \\"#;if var\\\\...",
  "VariableKey (2, 4): \\"if\\"",
  "Space (4, 5): \\" \\"",
  "CommentOperatorStatement (5, 9): \\"var\\\\n\\"",
  "StatementObject<NameInstance> (5, 8): \\"var\\"",
  "Block (9, 23): \\"foo=\\\\\\"bar...",
  "Variable (9, 19): \\"foo=\\\\\\"bar...",
  "VariableKey (9, 12): \\"foo\\"",
  "SymbolEqual (12, 13): \\"=\\"",
  "VariableValue (13, 18): \\"\\\\\\"bar\\\\\\"\\"",
  "CommentOperator (19, 23): \\"#;fi\\"",
  "VariableKey (21, 23): \\"fi\\"",
  "SpaceNewLine (23, 24): \\"\\\\n\\"",
]
`;

exports[`DataSource parse statement should parse comment operator correctly 5`] = `
Array [
  "Block (0, 32): \\"#;if foo ...",
  "CommentOperator (0, 31): \\"#;if foo ...",
  "VariableKey (2, 4): \\"if\\"",
  "Space (4, 5): \\" \\"",
  "CommentOperatorStatement (5, 17): \\"foo === b...",
  "StatementObject<NameInstance> (5, 8): \\"foo\\"",
  "Space (8, 9): \\" \\"",
  "StatementObject<StrictEqualitySymbol> (9, 12): \\"===\\"",
  "Space (12, 13): \\" \\"",
  "StatementObject<NameInstance> (13, 16): \\"bar\\"",
  "Block (17, 31): \\"foo=\\\\\\"bar...",
  "Variable (17, 27): \\"foo=\\\\\\"bar...",
  "VariableKey (17, 20): \\"foo\\"",
  "SymbolEqual (20, 21): \\"=\\"",
  "VariableValue (21, 26): \\"\\\\\\"bar\\\\\\"\\"",
  "CommentOperator (27, 31): \\"#;fi\\"",
  "VariableKey (29, 31): \\"fi\\"",
  "SpaceNewLine (31, 32): \\"\\\\n\\"",
]
`;

exports[`DataSource parse statement should parse statement with numbers 1`] = `
Object {
  "$type": "CommentOperatorStatement",
  "children": undefined,
  "end": 26,
  "pos": 0,
  "statements": Array [
    Object {
      "$type": "StatementObject",
      "children": undefined,
      "end": 8,
      "pos": 0,
      "type": "Number",
      "value": 34214324,
    },
    Object {
      "$type": "StatementObject",
      "children": undefined,
      "end": 12,
      "pos": 9,
      "type": "StrictEqualitySymbol",
      "value": undefined,
    },
    Object {
      "$type": "StatementObject",
      "children": undefined,
      "end": 24,
      "pos": 15,
      "type": "Number",
      "value": 22343.32,
    },
  ],
}
`;

exports[`DataSource shoud make a ast out 1`] = `
" Block (0, 60): \\"abc=def #...
  Variable (0, 20): \\"abc=def #...
    VariableKey (0, 3): \\"abc\\"
    SymbolEqual (3, 4): \\"=\\"
    VariableValue (4, 7): \\"def\\"
    Space (7, 8): \\" \\"
    CommentInline (8, 20): \\"# a comme...
      Space (9, 10): \\" \\"
  Variable (20, 28): \\"foo=bar\\\\n...
    VariableKey (20, 23): \\"foo\\"
    SymbolEqual (23, 24): \\"=\\"
    VariableValue (24, 27): \\"bar\\"
  SpaceNewLine (28, 34): \\"\\\\n    \\\\n\\"
  Variable (34, 60): \\"abc    =\\\\...
    VariableKey (34, 37): \\"abc\\"
    Space (37, 41): \\"    \\"
    SymbolEqual (41, 42): \\"=\\"
    VariableValue (42, 49): \\"\\\\\\"d\\\\\\\\\\\\\\"ef\\\\\\"\\"
    Space (49, 50): \\" \\"
    CommentInline (50, 60): \\"# comente...
      Space (51, 52): \\" \\""
`;

exports[`DataSource shoud make a ast with multiple blocks 1`] = `
" Block (0, 43): \\"C=C\\\\n#;if...
  Variable (0, 4): \\"C=C\\\\n\\"
    VariableKey (0, 1): \\"C\\"
    SymbolEqual (1, 2): \\"=\\"
    VariableValue (2, 3): \\"C\\"
  CommentOperator (4, 42): \\"#;if true...
    VariableKey (6, 8): \\"if\\"
    Space (8, 9): \\" \\"
    CommentOperatorStatement (9, 14): \\"true\\\\n\\"
      StatementObject<Boolean> (9, 13): \\"true\\"
    Block (14, 42): \\"a=a\\\\n#;if...
      Variable (14, 18): \\"a=a\\\\n\\"
        VariableKey (14, 15): \\"a\\"
        SymbolEqual (15, 16): \\"=\\"
        VariableValue (16, 17): \\"a\\"
      CommentOperator (18, 37): \\"#;if fals...
        VariableKey (20, 22): \\"if\\"
        Space (22, 23): \\" \\"
        CommentOperatorStatement (23, 29): \\"false\\\\n\\"
          StatementObject<Boolean> (23, 28): \\"false\\"
        Block (29, 37): \\"b=b\\\\n#;fi...
          Variable (29, 33): \\"b=b\\\\n\\"
            VariableKey (29, 30): \\"b\\"
            SymbolEqual (30, 31): \\"=\\"
            VariableValue (31, 32): \\"b\\"
          CommentOperator (33, 37): \\"#;fi\\"
            VariableKey (35, 37): \\"fi\\"
      SpaceNewLine (37, 38): \\"\\\\n\\"
      CommentOperator (38, 42): \\"#;fi\\"
        VariableKey (40, 42): \\"fi\\"
  SpaceNewLine (42, 43): \\"\\\\n\\""
`;

exports[`DataSource shoud make a ast with operators 1`] = `
" Block (0, 44): \\"FOO=BAR\\\\n...
  Variable (0, 8): \\"FOO=BAR\\\\n...
    VariableKey (0, 3): \\"FOO\\"
    SymbolEqual (3, 4): \\"=\\"
    VariableValue (4, 7): \\"BAR\\"
  CommentOperator (8, 34): \\"#; if tru...
    Space (10, 11): \\" \\"
    VariableKey (11, 13): \\"if\\"
    Space (13, 14): \\" \\"
    CommentOperatorStatement (14, 19): \\"true\\\\n\\"
      StatementObject<Boolean> (14, 18): \\"true\\"
    Block (19, 34): \\"aaa=\\\\\\"bbb...
      Variable (19, 29): \\"aaa=\\\\\\"bbb...
        VariableKey (19, 22): \\"aaa\\"
        SymbolEqual (22, 23): \\"=\\"
        VariableValue (23, 28): \\"\\\\\\"bbb\\\\\\"\\"
      CommentOperator (29, 34): \\"#; fi\\"
        Space (31, 32): \\" \\"
        VariableKey (32, 34): \\"fi\\"
  SpaceNewLine (34, 36): \\"\\\\n\\\\n\\"
  Variable (36, 44): \\"ccc=ddd\\\\n...
    VariableKey (36, 39): \\"ccc\\"
    SymbolEqual (39, 40): \\"=\\"
    VariableValue (40, 43): \\"ddd\\""
`;
